### PR TITLE
[NPM] Caching Pod information to compare against desired state

### PR DESCRIPTION
**Pull request by vakalapa**
This PR includes below changes:
1) Adding Caching ability for updated Pods to help relieve pressure on update events
2) optimizing named ports code path for uniformity
3) Removing generic PodMap and maintaining a NameSpace specific podmap
4) ResourceVersions checks for Namespaces, Pods and Network Policies

### CODE REVIEW

File: npm/pod.go [hunk 1]
```diff
@@ -29,36 +86,99 @@ func isInvalidPodUpdate(oldPodObj, newPodObj *corev1.Pod) (isInvalidUpdate bool)
 	isInvalidUpdate = oldPodObj.ObjectMeta.Namespace == newPodObj.ObjectMeta.Namespace &&
 		oldPodObj.ObjectMeta.Name == newPodObj.ObjectMeta.Name &&
 		oldPodObj.Status.Phase == newPodObj.Status.Phase &&
-		oldPodObj.Status.PodIP == newPodObj.Status.PodIP &&
```
Comment (vakalapa) [inline npm/pod.go:32]: Put this back

File: npm/util/const.go [hunk 2]
```diff
@@ -116,6 +116,7 @@ const (
 
 	IpsetUDPFlag  string = "udp:"
 	IpsetSCTPFlag string = "sctp:"
+	IpsetTCPFlag  string = "tcp:"
```
Comment (matmerr) [inline npm/util/const.go:119]: yeah how was this even missing :D

File: npm/namespace.go [hunk 3]
```diff
@@ -9,21 +9,21 @@
...
-	name           string
-	labelsMap      map[string]string
-	setMap         map[string]string
-	podMap         map[types.UID]*corev1.Pod
-	rawNpMap       map[string]*networkingv1.NetworkPolicy
-	processedNpMap map[string]*networkingv1.NetworkPolicy
-	ipsMgr         *ipsm.IpsetManager
-	iptMgr         *iptm.IptablesManager
+	name            string
+	labelsMap       map[string]string
```
Comment (neaggarwMS) [inline npm/namespace.go:19]: >string [](start = 21, length = 6)

nit: can you also add comment for what key/value represents for other maps

File: npm/namespace.go [hunk 4]
```diff
@@ -32,16 +32,24 @@
...
 		setMap:         make(map[string]string),
-		podMap:         make(map[types.UID]*corev1.Pod),
+		podMap:         make(map[string]*npmPod),
 		rawNpMap:       make(map[string]*networkingv1.NetworkPolicy),
 		processedNpMap: make(map[string]*networkingv1.NetworkPolicy),
 		ipsMgr:         ipsm.NewIpsetManager(),
 		iptMgr:         iptm.NewIptablesManager(),
+		// resource version is converted to uint64
+		// so make sure it is initialized to "0"
+		resourceVersion: "0",
```
Comment (neaggarwMS) [inline npm/namespace.go]: >resourceVersion [](start = 2, length = 15)

As discussed offline, lets store this as an int

File: npm/namespace.go [hunk 5]
```diff
@@ -56,10 +64,18 @@
...
+	if !exists {
+		return false
+	}
+
+	if isSamePolicy(np, npObj) {
+		return true
+	}
+
+	if !util.CompareResourceVersions(np.ObjectMeta.ResourceVersion, npObj.ObjectMeta.ResourceVersion) {
+		log.Logf("Cached Network Policy has larger ResourceVersion number than new Obj of %s\n", npObj.ObjectMeta.Name)
```
Comment (neaggarwMS) [inline npm/namespace.go:77]: >Cached Network Policy has larger ResourceVersion number than new Obj of [](start = 12, length = 71)

For debugging, can we also log version numbers from both cached and new object

File: npm/namespace.go [hunk 6]
```diff
@@ -56,10 +64,18 @@
...
+	np, exists := ns.rawNpMap[npObj.ObjectMeta.Name]
+	if !exists {
+		return false
+	}
+
+	if isSamePolicy(np, npObj) {
+		return true
+	}
+
+	if !util.CompareResourceVersions(np.ObjectMeta.ResourceVersion, npObj.ObjectMeta.ResourceVersion) {
```
Comment (neaggarwMS) [inline npm/namespace.go:76]: Shouldnt this check be done before isSamePolicy?

File: npm/namespace.go [hunk 7]
```diff
@@ -198,46 +215,41 @@
 		return nil
 	}
 
+	if !util.CompareUintResourceVersions(
+		curNsObj.resourceVersion,
+		util.ParseResourceVersion(newNsObj.ObjectMeta.ResourceVersion),
+	) {
```
Comment (neaggarwMS) [inline npm/namespace.go:221]: >) { [](start = 1, length = 3)

nit: move this to above line

File: npm/namespace.go [hunk 8]
```diff
@@ -198,46 +215,41 @@
...
+		log.Logf("Cached NameSpace has larger ResourceVersion number than new Obj of %s\n", oldNsNs)
```
Comment (neaggarwMS) [inline npm/namespace.go:222]: Same as above, add Version number as well from cached and new object list

**Comment by neaggarwMS** [comment]
> 	if oldNsNs != newNsNs {

Shouldnt we check resourceVersion before this?

---
Refers to: npm/namespace.go:182 in 62af54d. [](commit_id = 62af54dc9a123b0b819bdc88dc1a0e45b49b0697, deletion_comment = False)

File: npm/namespace.go [hunk 9]
```diff
@@ -155,8 +172,8 @@
 	}
 
 	var err error
-	oldNsNs, oldNsLabel := "ns-"+oldNsObj.ObjectMeta.Name, oldNsObj.ObjectMeta.Labels
-	newNsNs, newNsLabel := "ns-"+newNsObj.ObjectMeta.Name, newNsObj.ObjectMeta.Labels
+	oldNsNs, oldNsLabel := util.GetNSNameWithPrefix(oldNsObj.ObjectMeta.Name), oldNsObj.ObjectMeta.Labels
```
Comment (neaggarwMS) [inline npm/namespace.go:175]: We should pick the oldNSObj from cache

**Comment by neaggarwMS** [comment]
> 		if err = npMgr.DeleteNamespace(oldNsObj); err != nil {

Similarly delete all what we have in cache

---
Refers to: npm/namespace.go:183 in 62af54d. [](commit_id = 62af54dc9a123b0b819bdc88dc1a0e45b49b0697, deletion_comment = False)

**Comment by neaggarwMS** [comment]
> 	if oldNsNs != newNsNs {

Is this check possible? I thought Resource Names are immutable

---
Refers to: npm/namespace.go:182 in 62af54d. [](commit_id = 62af54dc9a123b0b819bdc88dc1a0e45b49b0697, deletion_comment = False)

**Review by neaggarwMS — DISMISSED** [review]
:clock1:

**Comment by neaggarwMS** [comment]
> 	}

Shouldnt this check be done after IsHostNetwork?

---
Refers to: npm/pod.go:192 in 62af54d. [](commit_id = 62af54dc9a123b0b819bdc88dc1a0e45b49b0697, deletion_comment = False)

File: npm/pod.go [hunk 10]
```diff
@@ -67,187 +193,266 @@
...
+	}
+
+	if isInvalidPodUpdate(getPodObjFromNpmObj(cachedPodObj), newPodObj) {
 		return nil
 	}
 
 	// today K8s does not allow updating HostNetwork flag for an existing Pod. So NPM can safely
 	// check on the oldPodObj for hostNework value
-	if isHostNetworkPod(oldPodObj) {
+	if isHostNetworkPod(newPodObj) {
```
Comment (neaggarwMS) [inline npm/pod.go:273]: Again we should avoid deep comparisons if pod is a hostnetwork pod

File: npm/pod.go [hunk 11]
```diff
@@ -67,187 +193,266 @@
...
+		}
+	}
+
+	cachedPodObj, exists := npMgr.nsMap[newPodObjNs].podMap[string(newPodObj.ObjectMeta.UID)]
+	if !exists {
+		if addErr := npMgr.AddPod(newPodObj); addErr != nil {
+			log.Errorf("Error: failed to add pod during update with error %+v", addErr)
+		}
+		return nil
+	}
```
Comment (neaggarwMS) [inline npm/pod.go:265]: This should be added only after we have validated host networkpod changes

File: npm/pod.go [hunk 12]
```diff
@@ -67,187 +193,266 @@
...
 		return nil
 	}
+	// We are assuming that FAILED to RUNNING pod will send an update
+	if newPodObj.Status.Phase == v1.PodSucceeded || newPodObj.Status.Phase == v1.PodFailed {
+		if delErr := npMgr.DeletePod(newPodObj); delErr != nil {
+			log.Errorf("Error: failed to add pod during update with error %+v", delErr)
+		}
+
+		return nil
+	}
```
Comment (neaggarwMS) [inline npm/pod.go:287]: Again, this should be done only after we compare the Resource versions

**Review by neaggarwMS — CHANGES_REQUESTED** [review]
:clock1:

File: npm/namespace.go (hunk 9 above)
Comment (vakalapa) [reply npm/namespace.go:175]: As discussed, we will creating a new Work item for these changes

File: npm/util/util.go [hunk 13]
```diff
@@ -238,3 +252,42 @@ func DropEmptyFields(s []string) []string {
...
+
+	return false
+}
+
+// ParseResourceVersion get uint64 version of ResourceVersion
+func ParseResourceVersion(rv string) uint64 {
+	if rv == "" {
+		return 0
+	}
+	rvInt, err := strconv.ParseUint(rv, 10, 64)
```
Comment (matmerr) [inline npm/util/util.go:287]: should potentially store resource version as uint if we're comparing every time, can take that in a later change

**Review by matmerr — APPROVED** [review]

**Review by neaggarwMS — APPROVED** [review]
:shipit:

**Pull request merged**